src/bsoncxx/types/value.cpp

```diff
@@ -22,14 +22,19 @@
 #include <bsoncxx/config/prelude.hpp>
 
 #include <bsoncxx/json.hpp>
+#include <bsoncxx/stdx/optional.hpp>
 
 namespace bsoncxx {
 BSONCXX_INLINE_NAMESPACE_BEGIN
 namespace types {
 
+// The default constructor does not initialize the value union.
+value::value() : _type(stdx::nullopt) {
+}
+
 #define BSONCXX_ENUM(name, val)                                                 \
     value::value(b_##name value)                                                \
-        : _type(static_cast<bsoncxx::type>(val)), _b_##name(std::move(value)) { \
+        : _type(stdx::optional<bsoncxx::type>(static_cast<bsoncxx::type>(val))), _b_##name(std::move(value)) { \
```

Comment:
You should be able to just say
```
_type(static_cast<bsoncxx::type>(val))
```
and the compiler will construct the optional for you.  Optional has constructor overloads that take a T directly: http://en.cppreference.com/w/cpp/experimental/optional/optional